### PR TITLE
Round out Mut<T> api using lenses + auto-unwrapping

### PR DESCRIPTION
**Pull request by TheRawMeatball**
# Objective

- Allow going from `Mut<LargeStruct>` to `Mut<Field>`
- Simplify working with components made with the newtype pattern

## Solution

Add lenses, which allow for getting a `Mut<U>` from a `Mut<T>` where `T` contains a `U`, and add a "default lens" option to the `Component` trait, allowing for automatic unwrapping of newtypes

## Changelog

- Added `Lens` trait
- Added associated methods to Mut<T> for lensing
- Added `DefaultLens` associated type to `Component` for automatic lensing
- Added `Raw` and `RawMut` query params which bypass automatic lensing

### CODE REVIEW

File: crates/bevy_ecs/src/query/fetch.rs [hunk 1]
```diff
@@ -495,10 +496,19 @@ impl<'w, 's> Fetch<'w, 's> for EntityFetch {
     }
 }
 
+/// Equilavent to querying with `&T`, but ignores the default lens. Useful with generic code
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/fetch.rs:499]:
```suggestion
/// Equivalent to querying with `&T`, but ignores the default lens. Useful with generic code
```

File: crates/bevy_ecs/src/query/fetch.rs [hunk 2]
```diff
@@ -644,34 +656,43 @@ impl<'w, 's, T: Component> Fetch<'w, 's> for ReadFetch<T> {
...
     }
 
     #[inline]
     unsafe fn table_fetch(&mut self, table_row: usize) -> Self::Item {
-        &*self.table_components.as_ptr().add(table_row)
+        L::get(&*self.table_components.as_ptr().add(table_row))
     }
 }
 
+/// Equilavent to querying with `&mut T`, but ignores the default lens. Useful with generic code
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/fetch.rs:668]:
```suggestion
/// Equivalent to querying with `&mut T`, but ignores the default lens. Useful with generic code
```

File: crates/bevy_render/src/view/visibility/mod.rs [hunk 3]
```diff
@@ -124,7 +124,7 @@ pub fn calculate_bounds(
 }
 
 pub fn update_frusta<T: Component + CameraProjection + Send + Sync + 'static>(
-    mut views: Query<(&GlobalTransform, &T, &mut Frustum)>,
+    mut views: Query<(&GlobalTransform, Raw<T>, &mut Frustum)>,
```
Comment (alice-i-cecile) [inline crates/bevy_render/src/view/visibility/mod.rs:127]: The need for this sort of change makes me worried: I don't love that users will have to reason about this when trying to write generic systems.

**Comment by alice-i-cecile** [comment]
Interesting. My first impressions are:

- this needs way more docs, especially focused around explaining what a lens is and why users care. I think the most useful thing would be a series of progressively more advanced doc examples
- this seems like it's competing for design space with #4328, and needs an explanation in the PR description of why it's better (and we should consider reverting #4328 if we go this direction).
- this seems like it might be useful for enabling granular change detection, especially on things like `Assets`

File: crates/bevy_render/src/view/visibility/mod.rs (hunk 3 above)
Comment (TheRawMeatball) [reply crates/bevy_render/src/view/visibility/mod.rs:127]: Maybe, but it's a fairly mindless fix - just use the Raw and RawMut wrappers for directly generic components. I think ergonomic wins for the more common cases outweigh this slight loss here.

**Comment by TheRawMeatball** [comment]
This is similar to #4328, but distinct in an important way: This method has benefits even when working with newtypes that wrap primitives. As for #4328, it can also be used for resources which could be its main selling point.

**Comment by alice-i-cecile** [comment]
> As for https://github.com/bevyengine/bevy/pull/4328, it can also be used for resources which could be its main selling point.

This only holds until we have `#[derive(Resource)]` though 🧪

**Comment by alice-i-cecile** [comment]
Closed in favor of #6199.

**Pull request closed without merging**